p2p: always set nTime for self-advertisements

mzumsande opened the pull request:
This logic was recently changed in https://github.com/bitcoin/bitcoin/commit/0cfc0cd32239d3c08d2121e028b297022450b320 to overwrite `addrLocal` with the address they gave us when self-advertising to an inbound peer. But if we don't also change `nTime` again from the default `TIME_INIT`, our peer will not relay our advertised address any further.

mzumsande committed:
p2p: always set nTime for self-advertisements

If we self-advertised to an inbound peer with the address they gave us,
nTime was left default-initialized, so that our peer wouldn't relay it
any further along.

mzumsande commented:
ping @vasild

naumenkogs commented:
utACK 65749ee3933898b54a77443e647ed5c1d12e0f99
good catch. could be a good functional test?

mzumsande commented:
> good catch. could be a good functional test?

I think that would be hard, because the local addresses in the functional tests are unroutable, so nodes don't self-advertise there. There is even the workaround of the `-addrmantest` parameter used in `p2p_node_network_limited.py` for this, but that just returns a fixed addr so the code changed here isn't executed.
However, the unit test `net_tests/get_local_addr_for_peer_port` covers this (and would fail without the update in this PR). Before, it would just check that the expected and actual CAddress have a default-initialized nTime.

vasild reviewed (commented):
Concept ACK

The proficiency of the engineers is just astonishing in this community (@mzumsande)! How the hell did you catch this?

mzumsande commented on src/test/net_tests.cpp:
```diff
@@ -678,7 +678,10 @@ BOOST_AUTO_TEST_CASE(get_local_addr_for_peer_port)
     in_addr peer_us_addr;
     peer_us_addr.s_addr = htonl(0x02030405);
-    const CAddress peer_us{CService{peer_us_addr, 20002}, NODE_NETWORK};
+    CAddress peer_us{CService{peer_us_addr, 20002}, NODE_NETWORK};
+    int64_t current_time{GetAdjustedTime()};
+    SetMockTime(current_time);
+    peer_us.nTime = current_time;
```
Wouldn't it be better to have the test compare the entire advertised `CAddress`, especially with the difficulties of writing functional test for self-advertisements? 
I know it is not the main intention for this test, but I think it is nice to have this tested on the side, especially with the footgunny `CAddress(CService, ServiceFlags)` ctor that is used so frequently, probably to avoid `int64_t`  / `uint32_t` compiler warnings when inserting a `int64_t` time into the full ctor with `nTime`.
So maybe it would be better to instead also raise the earlier check to the Level of `CAddress`?

vasild replied:
Alright, it is good to test more, even though this test is targeted at testing the ports stuff. IMO it is better for one test to test one thing, so maybe create a new test that just calls `GetLocalAddrForPeer()` without the ports complications from the `get_local_addr_for_peer_port` test. Anyway, up to you. If the existent test is to be modified, then I guess something like this should suffice:

```diff
     m_node.args->ForceSetArg("-bind", strprintf("3.4.5.6:%u", bind_port));
 
+    const uint32_t current_time = static_cast<uint32_t>(GetAdjustedTime());
+
+    SetMockTime(current_time);
+
     // Our address:port as seen from the peer, completely different from the above.
     in_addr peer_us_addr;
     peer_us_addr.s_addr = htonl(0x02030405);
-    const CAddress peer_us{CService{peer_us_addr, 20002}, NODE_NETWORK};
+    const CAddress peer_us{CService{peer_us_addr, 20002}, NODE_NETWORK, current_time};
 
     // Create a peer with a routable IPv4 address (outbound).
```

mzumsande replied:
I took your suggestion to modify, and also changed the check for `expected` to the level of `CAddress`. I think it's natural for a unit test of a function that returns an object to check the entire object is as expected, and not just parts of it, especially if it's just two more fields and a (thorough) separate test that would have caught this issue would have required to replicate some of the port stuff in order to also get to the branch where nTime was left default-initialized.

mzumsande commented on src/net.cpp:
```diff
@@ -251,6 +251,7 @@ std::optional<CAddress> GetLocalAddrForPeer(CNode *pnode)
             // For inbound connections, assume both the address and the port
             // as seen from the peer.
             addrLocal = CAddress{pnode->GetAddrLocal(), addrLocal.nServices};
+            addrLocal.nTime = GetAdjustedTime();
```
Good suggestion, will do with the next push.

mzumsande commented:
> The proficiency of the engineers is just astonishing in this community (@mzumsande)! How the hell did you catch this?

Thank you! Pretty much by chance - I was reviewing #25303 and while looking at this [commit msg](https://github.com/bitcoin/bitcoin/pull/25303/commits/00001e57fe74c061aa9cbc72b07252335cb566e0) I was wondering if it was possible to somehow have addresses with default-initialised nTime in addrman or addr relay  - that's how I stumbled upon this.

mzumsande commented:
[65749ee ](https://github.com/bitcoin/bitcoin/commit/65749ee3933898b54a77443e647ed5c1d12e0f99) -> [99b9e5f](https://github.com/bitcoin/bitcoin/commit/99b9e5f3a9fa29bbc1e45fc958470fbcc207ef23)
Addressed feedback by @vasild (thanks!)

vasild reviewed (approved):
ACK 99b9e5f3a9fa29bbc1e45fc958470fbcc207ef23

Thank you!

naumenkogs commented:
ACK 99b9e5f3a9fa29bbc1e45fc958470fbcc207ef23

laanwj commented:
Code review ACK 99b9e5f3a9fa29bbc1e45fc958470fbcc207ef23
Great catch!

mzumsande commented:
Maybe consider backporting to 23.x if this meets the bar.
